### PR TITLE
Add support for some JUnit Platform selectors

### LINKED ISSUE #517: Is there any reason to drop support for various selectors?
I was satisfied with Spek 1.x to use IntelliJ test class execution feature which has the following icon
image
But Spek 2.x executes all test classes despite of class selection and I found where this difference come from.

[1.x SpekTestEngine][1.x] awares JUnit's various selectors, but [2.x][2.x] detects only two kinds of selectors. Is there any reason?

[1.x]: https://github.com/spekframework/spek/blob/83da1920977eeb35af7c9487d80c28e207c43f57/spek-junit-platform-engine/src/main/kotlin/org/jetbrains/spek/engine/SpekTestEngine.kt#L70
[2.x]: https://github.com/spekframework/spek/blob/2.x/spek-runner/junit5/src/main/kotlin/org/spekframework/spek2/junit/SpekTestEngine.kt#L29

### LINKED ISSUE #519: Unable to run an individual JUnit test when using spek 2
I have a project that uses both junit 5 and spek 2 tests, but when I run an individual junit test, the spek tests execute as well, regardless of having the spek framework plugin (2.0.0-rc.1.150) installed or not.

screenshot 2018-10-23 at 10 10 15
screenshot 2018-10-23 at 10 10 56

The project uses gradle 4.10.2 and the configuration looks like this:
```
buildscript {
    repositories {
        jcenter()
    }
    dependencies {
        classpath "org.jetbrains.kotlin:kotlin-gradle-plugin:1.2.71"
    }
}

apply plugin: 'kotlin'

repositories {
    mavenCentral()
    jcenter()
}

dependencies {
    implementation "org.jetbrains.kotlin:kotlin-stdlib-jdk8"
    implementation "org.jetbrains.kotlin:kotlin-reflect"

    testImplementation 'org.junit.jupiter:junit-jupiter-api:5.3.1'
    testRuntimeOnly 'org.junit.jupiter:junit-jupiter-engine:5.3.1'
    testImplementation 'org.assertj:assertj-core'
  
    testImplementation ('org.spekframework.spek2:spek-dsl-jvm:2.0.0-rc.1')  {
        exclude group: 'org.jetbrains.kotlin'
    }
    testRuntimeOnly ('org.spekframework.spek2:spek-runner-junit5:2.0.0-rc.1') {
        exclude group: 'org.junit.platform'
        exclude group: 'org.jetbrains.kotlin'
    }
}

test {
    useJUnitPlatform {
        includeEngines 'junit-jupiter', 'spek2'
    }
}
```

I'm using Intellij 2018.2.5 on MacOs Mojave.

### PR DESCRIPTION
**Pull request by raniejade**
Resolves #517, resolves #519

Tested it by adding another test under `integration-test:jvm` and running `./gradlew test --tests <package or class>`.

### CODE REVIEW

File: spek-runner/junit5/src/main/kotlin/org/spekframework/spek2/junit/SpekTestEngine.kt [hunk 1]
```diff
@@ -31,10 +35,34 @@ class SpekTestEngine : TestEngine {
...
+                PathBuilder().append(it.packageName)
+                    .build()
+            }
+
+        val filters = mutableListOf<Path>()
+
+        filters.addAll(classSelectors)
+        filters.addAll(packageSelectors)
+
+        // todo: empty filter should imply root
```
Comment (raniejade) [inline spek-runner/junit5/src/main/kotlin/org/spekframework/spek2/junit/SpekTestEngine.kt:71]: To minimize the scope of this change, will be addressing it in another PR. Basically just don't filter if list is empty.

**Review by vanniktech — APPROVED** [review]
Can we write a test for this, somehow?

**Comment by raniejade** [comment]
@vanniktech I've got an idea 👍

**Comment by raniejade** [comment]
Apologies for random commits, travis is acting up :/

File: .travis.yml [hunk 2]
```diff
@@ -3,15 +3,22 @@ sudo: false
...
 language: java
 jdk:
-  - oraclejdk10
+  - openjdk10
+
+matrix:
+  include:
+  - jdk: openjdk10
+    before_install:
+    - rm "${JAVA_HOME}/lib/security/cacerts"
```
Comment (artem-zinnatullin) [inline .travis.yml]: Gradle issues?

File: .travis.yml [hunk 3]
```diff
@@ -3,15 +3,22 @@ sudo: false
...
+    - ln -s /etc/ssl/certs/java/cacerts "${JAVA_HOME}/lib/security/cacerts"
 
 branches:
   only:
     - 1.x
     - 2.x
     - /^\d+\.\d+\.\d+(-\S*)?$/
 
-install: skip
+install: true
```
Comment (artem-zinnatullin) [inline .travis.yml]: I'd put something like `echo skip`, true is a command here which is confusing because you're actually disabling the install step by overriding it :)

File: spek-runner/junit5/src/main/kotlin/org/spekframework/spek2/junit/SpekTestEngine.kt [hunk 4]
```diff
@@ -31,10 +35,34 @@ class SpekTestEngine : TestEngine {
...
+                    .build()
+            }
+
+        val packageSelectors = discoveryRequest.getSelectorsByType(PackageSelector::class.java)
+            .map {
+                PathBuilder().append(it.packageName)
+                    .build()
+            }
+
+        val filters = mutableListOf<Path>()
```
Comment (artem-zinnatullin) [inline spek-runner/junit5/src/main/kotlin/org/spekframework/spek2/junit/SpekTestEngine.kt]: nit: should we maintain `linkedHashSet` here for future? Will allow removing potential duplicates

**Review by artem-zinnatullin — APPROVED** [review]

File: .travis.yml (hunk 2 above)
Comment (raniejade) [reply .travis.yml]: @artem-zinnatullin Yeah, any ideas how to fix? travis now rejects `oraclejdk10` 🤷‍♂️

File: .travis.yml (hunk 3 above)
Comment (raniejade) [reply .travis.yml]: TIL, thanks!

File: spek-runner/junit5/src/main/kotlin/org/spekframework/spek2/junit/SpekTestEngine.kt (hunk 4 above)
Comment (raniejade) [reply spek-runner/junit5/src/main/kotlin/org/spekframework/spek2/junit/SpekTestEngine.kt]: 👌

File: .travis.yml (hunk 3 above)
Comment (raniejade) [reply .travis.yml]: Hopefully this will fix the gradle issue as well.

File: .travis.yml (hunk 2 above)
Comment (artem-zinnatullin) [reply .travis.yml]: Nah, this is fine, I was just confirming why it's here

Here is more details if you're curious https://github.com/gradle/gradle/pull/6721

File: spek-runner/junit5/src/test/kotlin/org/spekframework/spek2/junit/RunnerTests.kt [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+            LauncherDiscoveryRequestBuilder.request()
+                .filters(EngineFilter.includeEngines("spek2"))
+                .selectors(DiscoverySelectors.selectPackage("testData.package1"))
+                .build(),
+            listener
+        )
+
+        val summary = listener.summary
+
+        assertThat(summary.testsSucceededCount, equalTo(3L))
```
Comment (raniejade) [inline spek-runner/junit5/src/test/kotlin/org/spekframework/spek2/junit/RunnerTests.kt:35]: only way to verify 🤷‍♂️

File: spek-runner/junit5/src/test/kotlin/org/spekframework/spek2/junit/RunnerTests.kt [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+            LauncherDiscoveryRequestBuilder.request()
+                .filters(EngineFilter.includeEngines("spek2"))
+                .selectors(DiscoverySelectors.selectClass("testData.package1.SpekTest1"))
+                .build(),
+            listener
+        )
+
+        val summary = listener.summary
+
+        assertThat(summary.testsSucceededCount, equalTo(1L))
```
Comment (raniejade) [inline spek-runner/junit5/src/test/kotlin/org/spekframework/spek2/junit/RunnerTests.kt:51]: same here.
